**Exercise: 11-4** If the quantity you want to predict is categorical, you can use multinomial logistic regression, which is implemented in StatsModels with a function called `mnlogit`. As an exercise, let’s use it to guess whether a woman is married, cohabitating, widowed, divorced, separated, or never married; in the NSFG dataset, marital status is encoded in a variable called `rmarital`.

Suppose you meet a woman who is 25 years old, white, and a high school graduate whose annual household income is about $45,000. What is the probability that she is married, cohabitating, etc?

In [8]:
import numpy as np
import pandas as pd

import thinkstats2
import thinkplot

import statsmodels.formula.api as smf



In [9]:
import first

live, firsts, others = first.MakeFrames()

import nsfg #importing the data

resp = nsfg.ReadFemResp()
resp.index = resp.caseid
join = live.join(resp, on='caseid', rsuffix='_r')  # join to combine variables from the preganancy and respondent tables.

join.numbabes.replace([97], np.nan, inplace=True)

# I used a nonlinear model of age.
join['age2'] = join.age_r**2

let’s use it to guess whether a woman is married, cohabitating, widowed, divorced, separated, or never married; in the NSFG dataset, marital status is encoded in a variable called rmarital.

In [6]:
live.rmarital.unique() # unique values for married, cohabitating, widowed, divorced, separated, or never married

array([1, 6, 4, 2, 5, 3], dtype=int64)

In [7]:
formula='rmarital ~ age_r + age2 + C(race) + totincr + educat'
model = smf.mnlogit(formula, data=join)
results = model.fit()
results.summary() 

Optimization terminated successfully.
         Current function value: 1.092083
         Iterations 8


<class 'statsmodels.iolib.summary.Summary'>
"""
                          MNLogit Regression Results                          
==============================================================================
Dep. Variable:               rmarital   No. Observations:                 9148
Model:                        MNLogit   Df Residuals:                     9113
Method:                           MLE   Df Model:                           30
Date:                Sun, 14 Feb 2021   Pseudo R-squ.:                  0.1661
Time:                        18:17:11   Log-Likelihood:                -9990.4
converged:                       True   LL-Null:                       -11981.
                                        LLR p-value:                     0.000
================================================================================
  rmarital=2       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        8.9153      0.792     11.251      0.000       7.362      10.468
C(race)[T.2]    -0.9260      0.087    -10.705      0.000      -1.096      -0.756
C(race)[T.3]    -0.6335      0.133     -4.747      0.000      -0.895      -0.372
age_r           -0.3567      0.050     -7.132      0.000      -0.455      -0.259
age2             0.0047      0.001      6.054      0.000       0.003       0.006
totincr         -0.1301      0.011    -11.475      0.000      -0.152      -0.108
educat          -0.1940      0.018    -10.534      0.000      -0.230      -0.158
--------------------------------------------------------------------------------
  rmarital=3       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        2.9927      2.970      1.007      0.314      -2.829       8.815
C(race)[T.2]    -0.3963      0.235     -1.685      0.092      -0.857       0.065
C(race)[T.3]     0.0650      0.336      0.194      0.846      -0.593       0.723
age_r           -0.3141      0.174     -1.806      0.071      -0.655       0.027
age2             0.0064      0.003      2.532      0.011       0.001       0.011
totincr         -0.3217      0.032    -10.135      0.000      -0.384      -0.259
educat          -0.1093      0.048     -2.266      0.023      -0.204      -0.015
--------------------------------------------------------------------------------
  rmarital=4       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -3.6475      1.193     -3.059      0.002      -5.985      -1.310
C(race)[T.2]    -0.3303      0.091     -3.641      0.000      -0.508      -0.153
C(race)[T.3]    -0.8227      0.170     -4.853      0.000      -1.155      -0.490
age_r            0.1238      0.070      1.763      0.078      -0.014       0.261
age2            -0.0008      0.001     -0.814      0.416      -0.003       0.001
totincr         -0.2288      0.011    -20.058      0.000      -0.251      -0.206
educat           0.0661      0.016      4.015      0.000       0.034       0.098
--------------------------------------------------------------------------------
  rmarital=5       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -2.3978      1.269     -1.889      0.059      -4.886       0.090
C(race)[T.2]    -1.0493      0.101    -10.366      0.000      -1.248      -0.851
C(race)[T.3]    -0.6065      0.154     -3.937      0.000      -0.908      -0.305
age_r            0.2084      0.077      2.699      0.007       0.057       0.360
age2            -0.0030      0.001     -2.619      0.009      -0.005      -0.001
totincr         -0.2900      0.014    -20.314      0.000      -0.318      -0.262
educat          -0.0176      0.021     -0.835      0.404      -0.059       0.024
-

Now we can predict the family status of a woman that

age_r = 25 years old,  
race = 2 white,  
totincr= 11 income exceeds $45,000,  
education =12 high school graduate ,   


In [10]:
columns = ['age_r', 'age2', 'race', 'totincr', 'educat']
new = pd.DataFrame([[25, 25**2, 2, 11, 12]], columns=columns)
results.predict(new)

0         1         2         3         4         5
0  0.745689  0.128291  0.001624  0.032801  0.021887  0.069708


#### This person has a 75% chance of being currently married, 
#### a 13% chance of being "not married but living with opposite sex partner", 
#### 1% chance of being widowed , etc.